src/drivers/mediatek/mt8195/afe-memif.c

```diff
@@ -0,0 +1,725 @@
+		memif->memif_id = channel;
+		dma_chan_set_data(&dma->chan[channel], memif);
+	}
+	return 0;
+
+out:
+	if (dma->chan) {
+		for (channel = 0; channel < dma->plat_data.channels; channel++)
+			rfree(dma_chan_get_data(&dma->chan[channel]));
+		rfree(dma->chan);
+		dma->chan = NULL;
+	}
+
+	afe_remove(afe);
+
+	return -ENOMEM;
+}
+
+static int memif_remove(struct dma *dma)
+{
+	int channel;
+	struct mtk_base_afe *afe = afe_get();
+
+	if (!dma->chan) {
+		tr_err(&memif_tr, "MEMIF: remove called without probe, it's a no-op");
+		return 0;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		/* TODO Disable HW requests for this channel */
+
+		rfree(dma_chan_get_data(&dma->chan[channel]));
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	afe_remove(afe);
+
+	return 0;
+}
+
+static int memif_interrupt(struct dma_chan_data *channel, enum dma_irq_cmd cmd)
+{
+	struct mtk_base_afe *afe = afe_get();
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+	unsigned int sample_size =
+		((memif->format == SOF_IPC_FRAME_S16_LE) ? 2 : 4) * memif->channel;
+	unsigned int period = memif->period_size / sample_size;
+
+	if (channel->status == COMP_STATE_INIT)
+		return 0;
+
+	switch (cmd) {
+	case DMA_IRQ_STATUS_GET:
+		return afe_irq_get_status(afe, memif->irq_id);
+	case DMA_IRQ_CLEAR:
+		afe_irq_clear(afe, memif->irq_id);
+		return 0;
+	case DMA_IRQ_MASK:
+		afe_irq_disable(afe, memif->irq_id);
+	case DMA_IRQ_UNMASK:
+		afe_irq_config(afe, memif->irq_id, memif->rate, period);
+		afe_irq_enable(afe, memif->irq_id);
+		return 0;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/* TODO need convert number to platform MACRO */
+static int memif_get_attribute(struct dma *dma, uint32_t type, uint32_t *value)
+{
+	switch (type) {
+	case DMA_ATTR_BUFFER_ALIGNMENT:
+	case DMA_ATTR_COPY_ALIGNMENT:
+		*value = 4;
+		break;
+	case DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT:
+		*value = 16;
+		break;
+	case DMA_ATTR_BUFFER_PERIOD_COUNT:
+		*value = 4;
+		break;
+	default:
+		return -ENOENT;
+	}
+	return 0;
+}
+
+static int memif_get_data_size(struct dma_chan_data *channel, uint32_t *avail, uint32_t *free)
+{
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+	uint32_t hw_ptr;
+
+	/* update hw pointer from afe memif */
+	hw_ptr = afe_memif_get_cur_position(memif->afe, memif->memif_id);
+	tr_dbg(&memif_tr, "get_pos:0x%x, base:0x%x, dir:%d", hw_ptr, memif->dma_base,
+	       memif->direction);
+	tr_dbg(&memif_tr, "dma_size:%u, period_size:%d", memif->dma_size, memif->period_size);
+	hw_ptr -= memif->dma_base;
+
+	if (memif->direction)
+		memif->rptr = hw_ptr;
+	else
+		memif->wptr = hw_ptr;
+
+	*avail = (memif->wptr + memif->dma_size - memif->rptr) % memif->dma_size;
+	/* TODO, check if need alignment the available and free size to 1 period */
+	if (memif->direction)
+		*avail = (*avail / memif->period_size) +
+			 (uint32_t)((*avail % memif->period_size) != 0);
+	else
+		*avail = *avail / memif->period_size;
+	*avail *= memif->period_size;
```

Comment:
oh, so you just need to round up or down. Can `period_size` be not a power of 2? If not, you can use `ALIGN_UP()` and `ALIGN_DOWN()`